### PR TITLE
Create an insertion_sort transformer and incorporate it into the transpiling protocol

### LINKED ISSUE #6771: Create a transformer that reorders operations if they commute
**Is your feature request related to a use case or problem? Please describe.**
When `cirq.optimize_for_target_gateset(, gateset=cirq.CZTargetGateset(preserve_moment_structure=False))` runs on `CZ01 CZ12 CZ01` it returns its input without realizing that the output should be CZ12. This is because there is no transformer that reroders operations. A basic one would do version of `insertion_sort` where two operations are swable only if they commute.

A sketch of this looks like
```py3
@cirq.transformer(add_deep_support=True)
def insertion_sort(
    circuit: 'cirq.AbstractCircuit',
    *,
    context: Optional['cirq.TransformerContext'] = None,
) -> 'cirq.Circuit':
    ops = []
    for op in circuit.all_operations():
        st = []
        while len(ops) and op.qubits < ops[-1].qubits and cirq.commutes(ops[-1], op):
            st.append(ops.pop())
        ops.append(op)
        ops.extend(st[::-1])
    return cirq.Circuit(ops)
```

After implementing this transformer, we need to add it to the list of transformers applied during compilation where appropiate (i.e. `preserve_moment_structure=False`)

**What is the urgency from your perspective for this issue? Is it blocking important work?**

P1 - I need this no later than the next release (end of quarter)

### PR DESCRIPTION
**Pull request by NoureldinYosri**
Running the compilation workflow with the following command should produce circuits with a very short depth

```py3
cirq.optimize_for_target_gateset(circuit, gateset=cirq.CZTargetGateset(preserve_moment_structure=False, reorder_operations=True), max_num_passes=None)
```

cc: @ikd-sci 

fixes #6771

### CODE REVIEW

File: cirq-core/cirq/transformers/insertion_sort.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    for op in circuit.all_operations():
+        st = []
+        while (
+            len(final_operations)
+            and op.qubits < final_operations[-1].qubits
+            and protocols.commutes(final_operations[-1], op, default=False)
+        ):
+            st.append(final_operations.pop())
+        final_operations.append(op)
+        final_operations.extend(st[::-1])
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/insertion_sort.py:48]: We have no guarantee op.qubits are sorted, for example, they are not for the CZ gate which is symmetric in q0, q1.
We can also do without a temporary list - for example
```suggestion
    final_operations: List['cirq.Operation'] = []
    sorted_qubits: Dict[int, List['cirq.Qid']] = {}
    for pos, op in enumerate(circuit.all_operations()):
        op_qubits = sorted_qubits[id(op)] = sorted(op.qubits)
        for tail_op in reversed(final_operations):
            tail_qubits = sorted_qubits[id(tail_op)]
            if op_qubits < tail_qubits and protocols.commutes(tail_op, op, default=False):
                pos -= 1
                continue
            break
        final_operations.insert(pos, op)
```

File: cirq-core/cirq/transformers/insertion_sort_test.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def test_insertion_sort():
+    c = cirq.Circuit(
+        cirq.CZ(cirq.q(2), cirq.q(1)),
+        cirq.CZ(cirq.q(2), cirq.q(4)),
+        cirq.CZ(cirq.q(0), cirq.q(1)),
+        cirq.CZ(cirq.q(2), cirq.q(1)),
+    )
+    got = cirq.transformers.insertion_sort_transformer(c)
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/insertion_sort_test.py:26]: Nit - how about renaming `got` --> `c_sorted` for descriptive naming?

Same in cirq-core/cirq/transformers/target_gatesets/cz_gateset_test.py.

File: cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py [hunk 3]
```diff
@@ -146,11 +156,15 @@ def _intermediate_result_tag(self) -> Hashable:
     @property
     def preprocess_transformers(self) -> List['cirq.TRANSFORMER']:
         """List of transformers which should be run before decomposing individual operations."""
+        rorder_transfomers = (
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py:159]: Nit
```suggestion
        reorder_transfomers = (
```

File: cirq-core/cirq/transformers/target_gatesets/cz_gateset_test.py [hunk 4]
```diff
@@ -319,3 +319,20 @@ class UnsupportedExample(cirq.testing.TwoQubitGate):
...
+    got = cirq.optimize_for_target_gateset(
+        c,
+        gateset=cirq.CZTargetGateset(preserve_moment_structure=False, reorder_operations=True),
+        max_num_passes=1,
+    )
+    assert got == cirq.Circuit(cirq.CZ(cirq.q(1), cirq.q(2)))
+
+
+def test_reorder_operations_and_preserve_moment_structure_raises():
+    with pytest.raises(ValueError, match='True'):
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/target_gatesets/cz_gateset_test.py:337]: Nit - perhaps we should be more narrow about the expected text -
```suggestion
    with pytest.raises(
        ValueError, match='reorder_operations and preserve_moment_structure can not both be True'
    ):
```

**Review by pavoljuhas — CHANGES_REQUESTED** [review]
We need to order by sorted qubits, otherwise LGTM with a few minor comments.

Thanks for adding this!

File: cirq-core/cirq/transformers/insertion_sort.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            j
+            and _id(sorted_operations[j]) < _id(sorted_operations[j - 1])
+            and protocols.commutes(sorted_operations[j], sorted_operations[j - 1], default=False)
+        ):
+            sorted_operations[j], sorted_operations[j - 1] = (
+                sorted_operations[j - 1],
+                sorted_operations[j],
+            )
+            j -= 1
+    return circuits.Circuit(sorted_operations)
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/insertion_sort.py:56]: Please reconsider the implementation I suggested in https://github.com/quantumlib/Cirq/pull/6776#discussion_r1805681136.  Unlike here, the op.qubits are sorted just once and cached, and the output list is modified just once per iteration - after finding the correct insertion position.

**Review by pavoljuhas — COMMENTED** [review]
please consider the suggested implementation.

**Comment by pavoljuhas** [comment]
> I'm willing to bet that my implementation is faster. the most expensive operation is not the qubit comparison but the call to commutes

Challenge accepted :-)
It actually turned out to be the comparison of Qid-s , even when sorted, that was taking up the most time.
The runtime got about twice faster after mapping qubits to integers; another optimization was to check for operations with disjoint qubits before calling protocols.commutes - see https://github.com/pavoljuhas/Cirq/commit/1ae465e118520e376de2820a1469a150459f7973.

Here is a benchmark code to be saved as `bench_isort.ipy` -
```
import cirq
moment = cirq.Moment(cirq.CZ(cirq.q(i + 1), cirq.q(i)) for i in reversed(range(0, 64, 2)))
circuit = cirq.Circuit(100 * [moment])
%timeit -n1 cirq.transformers.insertion_sort_transformer(circuit)
```

and here is the outcome for the PR head and the optimized version:

```bash
# current head of PR #6776
$ git checkout a8277a63bcdf69f20a0a81b182a2643220bb23ff
$ ipython bench_isort.ipy

14.1 s ± 93.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# proposed change, https://github.com/pavoljuhas/Cirq/tree/amend-6776
$ git fetch https://github.com/pavoljuhas/Cirq.git amend-6776
$ git checkout 1ae465e118520e376de2820a1469a150459f7973
$ ipython bench_isort.ipy

5.59 s ± 72.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by NoureldinYosri** [comment]
after my last commit, it takes 0.5 a second on my machine
image

File: cirq-core/cirq/transformers/insertion_sort.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+        def _msk(qs: Tuple['cirq.Qid', ...]) -> int:
+            msk = 0
+            for q in qs:
+                msk |= 1 << q_index[q]
+            return msk
+
+        operations_with_info: Union[
+            List[Tuple['cirq.Operation', int, int]], List[Tuple['cirq.Operation', int, FrozenSet]]
+        ] = [
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/insertion_sort.py:56]: I don't think it is worthwhile to introduce this branching and a duality of type of the third tuple element which can be `int` or `FrozenSet`, both of which are later used with the `&` operator.  The extra human overhead in reading and understanding the code and a higher likelihood of bugs in a more complicated implementation is IMO not worth the time saved on a special case benchmark.

On my host - after upgrading to Python 3.13.0 the original benchmark times are slightly better on your side when it can use bitmask instead of set -
```
# PR 6776 at 9c31fe6f
491 ms ± 21.1 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
# https://github.com/pavoljuhas/Cirq/tree/amend-6776 at ee1e264e
593 ms ± 28.1 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
```
However, after increasing the number of qubits to 100, to use the second branch below, the times were
```
# PR 6776 at 9c31fe6f
1.67 s ± 132 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
# https://github.com/pavoljuhas/Cirq/tree/amend-6776 at ee1e264e
1.49 s ± 57.3 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
```

To have a more thorough check, I have tried another benchmark `bench_isort_z.ipy`, which needs to check commutation between CZ and Z gates -
```
# bench_isort_z.ipy
import cirq

moment_cz = cirq.Moment(cirq.CZ(cirq.q(i + 1), cirq.q(i)) for i in reversed(range(0, 64, 2)))
moment_z = cirq.Moment(cirq.Z(cirq.q(i)) for i in reversed(range(64)))
circuit = cirq.Circuit(20 * [moment_cz, moment_z])

%timeit -r10 cirq.transformers.insertion_sort_transformer(circuit)
```
For this one, the timings with this PR and the [amend-6776 branch](https://github.com/pavoljuhas/Cirq/commit/ee1e264e47673f5c1255136aacee61a6ef49bf83) show no significant difference and are both likely bound by the `protocols.commutes` calls:
```
# PR 6776 at 9c31fe6f
4.04 s ± 166 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
# https://github.com/pavoljuhas/Cirq/tree/amend-6776 at ee1e264e
3.99 s ± 96.6 ms per loop (mean ± std. dev. of 10 runs, 1 loop each)
```

**Review by pavoljuhas — CHANGES_REQUESTED** [review]
Please change to a non-branched implementation.
I put more info into an inline comment, but in brief if the transformer uses `protocols.commutes` there is no performance difference between this PR and my suggestion in https://github.com/pavoljuhas/Cirq/commit/ee1e264e47673f5c1255136aacee61a6ef49bf83.  The times are comparable also for the initial benchmark if circuits have 65 or more qubits.

Also please try to look at the code as someone who sees it for the first time.  It is important to have a good performance, but it is also important to have a simple, easy to follow code - this decreases the likelihood of errors and the maintenance toil further on.  :-)

**Comment by NoureldinYosri** [comment]
@pavoljuhas I copied your implementation

**Review by pavoljuhas — APPROVED** [review]
I have fixed corner-case of for zero-qubit gates and cleaned
up test parameterization.  Thank you Nour!
